NorthstarDLL/logging/crashhandler.cpp

```diff
@@ -1,377 +1,527 @@
+// Purpose: Destructor
+//-----------------------------------------------------------------------------
+CCrashHandler::~CCrashHandler()
 {
+	Shutdown();
+}
 
-	MODULEINFO crashedModuleInfo;
-	GetModuleInformation(GetCurrentProcess(), exc.crashedModule, &crashedModuleInfo, sizeof(crashedModuleInfo));
-
-	char crashedModuleFullName[MAX_PATH];
-	GetModuleFileNameExA(GetCurrentProcess(), exc.crashedModule, crashedModuleFullName, MAX_PATH);
-	char* crashedModuleName = strrchr(crashedModuleFullName, '\\') + 1;
+//-----------------------------------------------------------------------------
+// Purpose: Initilazes crash handler
+//-----------------------------------------------------------------------------
+void CCrashHandler::Init()
+{
+	m_hExceptionFilter = AddVectoredExceptionHandler(TRUE, ExceptionFilter);
+	m_bHasSetConsolehandler = SetConsoleCtrlHandler(ConsoleCtrlRoutine, TRUE);
+}
 
-	DWORD64 crashedModuleOffset = ((DWORD64)exc.exceptionRecord.ExceptionAddress) - ((DWORD64)crashedModuleInfo.lpBaseOfDll);
+//-----------------------------------------------------------------------------
+// Purpose: Shutdowns crash handler
+//-----------------------------------------------------------------------------
+void CCrashHandler::Shutdown()
+{
+	if (m_hExceptionFilter)
+	{
+		RemoveVectoredExceptionHandler(m_hExceptionFilter);
+		m_hExceptionFilter = nullptr;
+	}
 
-	PVOID framesToCapture[62];
-	int frames = RtlCaptureStackBackTrace(0, 62, framesToCapture, NULL);
-	bool haveSkippedErrorHandlingFrames = false;
+	if (m_bHasSetConsolehandler)
+	{
+		SetConsoleCtrlHandler(ConsoleCtrlRoutine, FALSE);
+	}
+}
 
-	for (int i = 0; i < frames; i++)
+//-----------------------------------------------------------------------------
+// Purpose: Sets the exception info
+//-----------------------------------------------------------------------------
+void CCrashHandler::SetExceptionInfos(EXCEPTION_POINTERS* pExceptionPointers)
+{
+	m_pExceptionInfos = pExceptionPointers;
+}
+//-----------------------------------------------------------------------------
+// Purpose: Sets the exception stirngs for message box
+//-----------------------------------------------------------------------------
+void CCrashHandler::SetCrashedModule()
+{
+	LPCSTR pCrashAddress = static_cast<LPCSTR>(m_pExceptionInfos->ExceptionRecord->ExceptionAddress);
+	HMODULE hCrashedModule;
+	if (!GetModuleHandleExA(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS, pCrashAddress, &hCrashedModule))
 	{
+		m_svCrashedModule = CRASHHANDLER_GETMODULEHANDLE_FAIL;
+		m_svCrashedOffset = "";
 
-		HMODULE backtraceModuleHandle;
-		GetModuleHandleExA(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS, static_cast<LPCSTR>(framesToCapture[i]), &backtraceModuleHandle);
+		wchar_t wszError[256];
+		// NOTE [Fifty]: MAKELANGID is deprecated
+		FormatMessageW(FORMAT_MESSAGE_FROM_SYSTEM, NULL, GetLastError(), MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), wszError, 255, NULL);
```

Comment:
Looking at the docs I think I should do error checking on this. To answer your question I very much do not know how this would act. I'll test when I get the chance though

Reply:
addressed in 688690d